integration-cli/docker_api_containers_unix_test.go

```diff
@@ -0,0 +1,49 @@
+// +build !windows
+
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+	"syscall"
+
+	"github.com/docker/docker/integration-cli/checker"
+	"github.com/docker/docker/integration-cli/request"
+	"github.com/docker/docker/pkg/archive"
+	"github.com/go-check/check"
+)
+
+func (s *DockerSuite) TestContainerAPIArchiveWithHardlink(c *check.C) {
+	testRequires(c, DaemonIsLinux)
+
+	containerID := makeTestContainer(c, testContainerOptions{addContent: true})
+
+	tmpDir := getTestDir(c, "test-cp-from-dir-with-hardlink")
+	defer os.RemoveAll(tmpDir)
+
+	srcPath := "/dir6"
+	dstPath := cpPath(tmpDir, "dir6_")
+	dstFile := cpPath(tmpDir, "dir6_/hardLinkFile1")
+	endpoint := fmt.Sprintf("/containers/%s/archive?path=%s", containerID, srcPath)
+	res, body, err := request.SockRequestRaw("GET", endpoint, nil, "application/x-tar", daemonHost())
+	c.Assert(err, checker.IsNil)
+	c.Assert(res.StatusCode, checker.Equals, http.StatusOK)
+
+	dstDir, dstBase := archive.SplitPathDirEntry(dstPath)
+	_, srcBase := archive.SplitPathDirEntry(srcPath)
+	copyArchive := archive.RebaseArchiveEntries(body, srcBase, dstBase)
```

Comment:
https://github.com/moby/moby/pull/33956 add the following three line, and the code is in RebaseArchiveEntries.
```
+++ b/pkg/archive/copy.go
@@ -332,6 +332,9 @@ func RebaseArchiveEntries(srcContent io.Reader, oldBase, newBase string) io.Read
                        }
 
                        hdr.Name = strings.Replace(hdr.Name, oldBase, newBase, 1)
+                       if hdr.Typeflag == tar.TypeLink {
+                               hdr.Linkname = strings.Replace(hdr.Linkname, oldBase, newBase, 1)
+                       }
 
                        if err = rebasedTar.WriteHeader(hdr); err != nil {
                                w.CloseWithError(err)
```